client/lib/backend_client.dart

```diff
@@ -78,28 +87,38 @@ class BackendClient {
     /// import 'package:example_dependency/types.dart' as dependencyTypes;
     /// import 'types.dart' as myTypes;
     ///
-    /// var firebase = RestClient(fb.FirebaseClient.anonymous(),
-    ///                           'https://example-org-42.firebaseio.com');
+    /// var firebaseClient = RestClient(fb.FirebaseClient.anonymous(),
+    ///                                 'https://example-org-42.firebaseio.com');
     /// var client = BackendClient('https://example.org',
-    ///                            firebase,
+    ///                            firebase: firebaseClient,
     ///                            typeRegistries: [myTypes.types(), dependencyTypes.types()]);
     /// ```
     ///
     BackendClient(String serverUrl,
-                  this._database,
-                  {List<dynamic> typeRegistries = const [],
-                   this.subscriptionKeepUpPeriod = _defaultSubscriptionKeepUpPeriod})
-            : _endpoint = HttpEndpoint(serverUrl) {
-        for (var registry in typeRegistries) {
-            theKnownTypes.register(registry);
+                 {FirebaseClient firebase,
+                  List<dynamic> typeRegistries = const [],
+                  this.subscriptionKeepUpPeriod = _defaultSubscriptionKeepUpPeriod,
+                  QueryMode queryMode = QueryMode.FIREBASE,
+                  Endpoints endpoints})
+            : _endpoint = HttpEndpoint(serverUrl),
+              _database = firebase,
+              _queryMode = queryMode {
+        ArgumentError.checkNotNull(serverUrl);
+        ArgumentError.checkNotNull(typeRegistries);
+        ArgumentError.checkNotNull(_queryMode);
+        if (_queryMode == QueryMode.FIREBASE && firebase == null) {
```

Comment:
Are you satisfied with these `if`s on the query mode the code? 

What if instead we created an interface-façade for this behaviour and two implementations, per each mode? Then we'll have just one `if` in the ctor, and mode-specific methods will be put into their own place instead of crowding the `BackendClient` altogether. 
Also, it will be much easier to append the client API with more modes once the time comes.